### PR TITLE
Added show_excerpt and max_words shortcode attributes

### PR DESCRIPTION
**Pull request by mikesetzer**
**Description**

- Added show_excerpt and max_words shortcode attributes to allow for the display of excerpts on the card layout
- Added css for new card excerpt
- Added shortcode attributes to the read me
- Configured plugin to retrieve post content and clean html and wordpress block tags, if no excerpt is found.

**Motivation and Context**

Increases the plugin display flexibility. 

**How Has This Been Tested?**

- Tested with negative, null, positive, and non-int values for max_words.
- Tested with and without excerpts.
- Tested on mobile.

**Types of changes**

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

**Checklist:**

- [x] My code follows the code style of this project.
- [x] My change requires an update to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by jmbarne3** [comment]
Hey Mike! Most of us are going to be out until Monday, so I'm going to set a reminder within our Slack to look this over then.

File: includes/ucf-post-list-common.php [hunk 1]
```diff
@@ -230,6 +230,41 @@ public static function get_image_or_fallback( $item, $size='large' ) {
...
+				} else {
+					$excerpt = implode(" ", $filtered_words);
+				}
+			} else {
+				$excerpt = implode (" ", $filtered_words);
+			}
+
+			return $excerpt;
+		}
+
```
Comment (jmbarne3) [inline includes/ucf-post-list-common.php:267]: You might be able to simplify this down using [`wp_trim_words`](https://developer.wordpress.org/reference/functions/wp_trim_words/).

Comment (mikesetzer) [reply includes/ucf-post-list-common.php]: Tested wp_trim_words but it would only filter html tags and leave shortcodes. Tried wp_trim_words on strings that had already been run through the preg_replace function used to strip shortcodes below.

$stripped_content = preg_replace("~(?:\[/?)[^/\]]+/?\]~s", '', $filtered_content);

**Comment by mikesetzer** [comment]
Two additional commits have been added that include a new layout for a material card design. Unsure if I should've closed and just resubmitted as a new pull request.

Material layout was tested on mobile, tablet, and desktop. Functionally the same as the card layout with excerpts, just a different design.

File: includes/ucf-post-list-common.php [hunk 2]
```diff
@@ -230,6 +230,44 @@ public static function get_image_or_fallback( $item, $size='large' ) {
...
+		/**
+		 * Retrieves the post excerpt and sets a character limit
+		 *
+		 * @author Mike Setzer
+		 * @since 2.0.8
+		 * @param object $item | object containing the WP Post
+		 * @param int $max_words | Maximum characters displayed from excerpt
+		 * @return string $excerpt | Excerpt text
+		 **/
+		public static function get_excerpt ( $item, $max_words) {
```
Comment (cjg89) [inline includes/ucf-post-list-common.php:242]: Like Jim noted, this function can be simplified a bit and updated to support existing excerpt filters with something like the following:

```php
public static function get_excerpt( $item, $max_words ) {
	$excerpt = '';

	$custom_filter = function( $l ) use ( $max_words ) {
		if ( $max_words > 0 ) {
				return $max_words;
		}
		return $l;
	};

	add_filter( 'excerpt_length', $custom_filter, 999 );
	$excerpt = wp_strip_all_tags( get_the_excerpt( $item ) );
	remove_filter( 'excerpt_length', $custom_filter, 999 );

	return $excerpt;
}
```

By using `get_the_excerpt()`, we ensure that developer-intended overrides to excerpts throughout the theme are applied (such as trailing ellipses modifications, the baseline excerpt length, etc); `get_the_excerpt()` will also grab a fallback value from the post content if an excerpt isn't explicitly defined.  The `add_filter` and `remove_filter` calls update and then reset the desired excerpt length on the fly.

If shortcode content is still a concern after trying this code, I'd recommend either adding shortcode parsing to excerpts **in your theme** (e.g. via `add_filter( 'the_excerpt', 'do_shortcode' );`), or add a similar function in your theme to strip the shortcode tags entirely; shortcode handling for excerpts probably falls outside of the scope of what the post list shortcode should be configured to do.

File: layouts/ucf-post-list-material.php [hunk 3]
```diff
@@ -0,0 +1,100 @@
+<?php
```
Comment (cjg89) [inline layouts/ucf-post-list-material.php:1]: A new shortcode layout falls outside of the scope of this PR--please remove the "material" layout addition in this PR and submit it as its own PR if you feel it's an appropriate inclusion in this plugin for everyone using it.  Thanks!

File: includes/ucf-post-list-config.php [hunk 4]
```diff
@@ -111,6 +112,15 @@ public static function get_options() {
 					'default' => 'Search'
 				) ),
 				'list_title'          => new UCF_Post_List_Option( 'list_title' ),
+				'show_excerpt'		  => new UCF_Post_List_Option( 'list_excerpt', array (
+					'default'		  => false,
+					'format_callback' => array( 'UCF_Post_List_Config', 'format_option_bool')
+				) ),
+				'max_words'			  => new UCF_Post_List_Option( 'max_words', array (
```
Comment (cjg89) [inline includes/ucf-post-list-config.php:119]: From a naming perspective, I think `excerpt_length` might be a little more clear here vs `max_words`.

File: layouts/ucf-post-list-card.php [hunk 5]
```diff
@@ -60,6 +65,9 @@ function ucf_post_list_display_card( $content, $posts, $atts ) {
 						<?php endif; ?>
 						<div class="ucf-post-list-card-block">
 							<h3 class="ucf-post-list-card-title"><?php echo $item->post_title; ?></h3>
+							<?php if( $atts['show_excerpt'] ) : ?>
+								<p class="ucf-post-list-excerpt-text ucf-post-list-card-text"><?php echo $item_excerpt; ?></p>
```
Comment (cjg89) [inline layouts/ucf-post-list-card.php:69]: Please use a `<div>` to wrap the excerpt here instead of a `<p>` tag--excerpts by default can be wrapped in `<p>` tags; we just override this behavior in the UCF WordPress Theme.  Using a `<div>` instead will be more compatible with other themes.

**Review by cjg89 — CHANGES_REQUESTED** [review]

File: layouts/ucf-post-list-card.php (hunk 5 above)
Comment (mikesetzer) [reply layouts/ucf-post-list-card.php:69]: I'm putting together a revision and will include that. However I've noticed the date field is in `<p>` tags: 
`<p class="ucf-post-list-card-text"><?php echo $date; ?></p>`

Would you like me to change that to `<div>` in my revision?

Comment (cjg89) [reply layouts/ucf-post-list-card.php:69]: The date being wrapped in a `<p>` is fine as-is, since `$date` should always be a date string with no additional formatting--but thanks for looking out 🙂

File: includes/ucf-post-list-config.php [hunk 6]
```diff
@@ -212,6 +221,9 @@ public static function get_options() {
 
 				// Tag
 				'tag'         => new UCF_Post_List_Option( 'tag' ),
+				'tag_name'    => new UCF_Post_List_Option( 'tag_name' , array(
```
Comment (cjg89) [inline includes/ucf-post-list-config.php:224]: Can this `tag_name` param be removed?  `tag_name` doesn't appear to be a supported parameter for `WP_Query`, and would need additional custom logic to work as intended.  https://developer.wordpress.org/reference/classes/wp_query/#tag-parameters

Comment (mikesetzer) [reply includes/ucf-post-list-config.php:224]: Just removed it. Was working on that on another branch and accidentally brought that over.

**Review by cjg89 — APPROVED** [review]

**Review by jmbarne3 — APPROVED** [review]
Nice set of changes. Cheers!

**Review by rjucf — APPROVED** [review]

**Review by cadie — APPROVED** [review]
